### PR TITLE
msglist: Make mark-as-read button listen to the Unreads model

### PR DESCRIPTION
**Pull request by chrisbobbe**
Instead of just listening to PerAccountStore and (via an ancestor widget) a MessageListView.

Issue #370 is relevant here -- "Maintain total unread counts efficiently". The button causes linear scans through Unreads each time it rebuilds, and ideally we'd avoid that, especially since now it'll rebuild in more cases than it was before. I think the additional cases aren't very many, though: I've thought of just one kind of Zulip event that would cause Unreads but not PerAccountStore and MessageListView to notify listeners, and I added a test exercising that.

### CODE REVIEW

File: test/widgets/message_list_test.dart [hunk 1]
```diff
@@ -1001,6 +1001,43 @@ void main() {
...
+        ],
+        mentions: [message.id],
+      );
+      await setupMessageListPage(tester,
+        narrow: MentionsNarrow(),
+        unreadMsgs: unreadMsgs,
+        // omit `message`; if present, MessageListView would notify listeners
+        messages: List.generate(300, (i) =>
+          eg.streamMessage(id: 950 + i, sender: eg.selfUser,
+            flags: [MessageFlag.read, MessageFlag.mentioned])),
```
Comment (gnprice) [inline test/widgets/message_list_test.dart:1244]: Ah, I see. IIUC, the point is that the message had been in the narrow, and an event removes it from the narrow; but it hadn't been one of the messages fetched, because it was too old. Is that right? I think that high-level strategy would be good to make explicit in the test's main comment above.

It looks like the same thing would happen with a channel or topic narrow if the message got moved out of the channel/topic, too.

File: test/widgets/message_list_test.dart [hunk 2]
```diff
@@ -1001,6 +1001,43 @@ void main() {
...
 
+    testWidgets('listens to Unreads model, not just PerAccountStore and MessageListView', (tester) async {
+      // Regression test for an edge case where the button wouldn't disappear
+      // when the narrow's unreads were cleared, because the button would only
+      // respond to notifications from PerAccountStore and MessageListView,
+      // not Unreads. To test this, we simulate an event that causes Unreads
+      // but not PerAccountStore and MessageListView to notify listeners,
+      // and check that the button responds.
+
+      final message = eg.streamMessage(flags: [MessageFlag.mentioned]);
```
Comment (gnprice) [inline test/widgets/message_list_test.dart]: Since the premise of this test scenario is that this message is older than all the messages that get fetched, it'd be good to give it a smaller message ID too in order to be consistent with that.

That or leave all the message IDs out ­— they're really only needed when the messages get constructed in the test out of order from when they were supposed to be sent.

**Comment by chrisbobbe** [comment]
> The new test exercises a scenario where the old code didn't behave right, so that's good to fix. But there was something else that prompted taking a look at this logic, right? ISTR a conversation a couple of weeks ago (before my vacation) where we were looking at `notifyListeners` calls in the context of a different change you were making. Remind me what the context was there?

I wonder if you're thinking about https://github.com/zulip/zulip-flutter/pull/1997#discussion_r2552051648 ?

The fix in this PR is something I hadn't noticed at the time. IIRC I spotted it (and #2026) when reviewing #2006.

**Comment by gnprice** [comment]
> I wonder if you're thinking about [#1997 (comment)](https://github.com/zulip/zulip-flutter/pull/1997#discussion_r2552051648) ?

No, what I recall is looking at a proposed change that added `notifyListeners` calls on PerAccountStore, for (some cases of?) some of the event types where we currently avoid those and instead notify only more-specific `ChangeNotifier`s. (Which is useful for the handful of high-frequency event types like MessageEvent and PresenceEvent.)

Anyway :shrug: I guess that was a coincidence. Not too surprising if there were just two unrelated reasons to be looking at this general area of logic in the span of a few weeks.

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

**Comment by gnprice** [comment]
Thanks! Looks good; merging.
